plugins/twbt-utils.cpp

```diff
@@ -0,0 +1,122 @@
+#include "Core.h"
+#include "Console.h"
+#include "Export.h"
+#include "PluginManager.h"
+#include "VersionInfo.h"
+#include "VTableInterpose.h"
+#include "LuaTools.h"
+
+#include "DataDefs.h"
+
+#include "df/viewscreen_dwarfmodest.h"
+#include "df/init.h"
+#include "df/renderer.h"
+#include "df/graphic.h"
+#include "df/enabler.h"
+#include "df/map_renderer.h"
+
+using std::string;
+using std::vector;
+using namespace DFHack;
+using namespace df::enums;
+
+DFHACK_PLUGIN("twbt-utils");
+REQUIRE_GLOBAL(window_x)
+REQUIRE_GLOBAL(window_y)
+REQUIRE_GLOBAL(window_z)
+REQUIRE_GLOBAL_NO_USE(gps)
+REQUIRE_GLOBAL_NO_USE(enabler)
+REQUIRE_GLOBAL_NO_USE(twbt_render_map)
+
+#ifdef WIN32
+    // On Windows there's no parameter pointing to the map_renderer structure
+    typedef void(_stdcall *RENDER_MAP)(int);
+
+    RENDER_MAP _render_map;
+
+    void render_map(){ _render_map(0); }
+#else
+REQUIRE_GLOBAL_NO_USE(map_renderer)
+
+    typedef void(*RENDER_MAP)(void*, int);
+
+    RENDER_MAP _render_map;
+
+    void render_map(){ _render_map(df::global::map_renderer,0); }
+#endif
+static int render_map_rect(lua_State* L)
+{
+    CoreSuspender suspender;
+
+    int x = luaL_checkint(L, 1);
+    int y = luaL_checkint(L, 2);
+    int z = luaL_checkint(L, 3);
+    int w = luaL_checkint(L, 4);
+    int h = luaL_checkint(L, 5);
+    //backup state
+    uint8_t *s = df::global::gps->screen;
+    int32_t win_h = df::global::gps->dimy;
+    int32_t was_x = *window_x;
+    int32_t was_y = *window_y;
+    int32_t was_z = *window_z;
+    int32_t gx = df::global::init->display.grid_x;
+    int32_t gy = df::global::init->display.grid_y;
+    df::global::init->display.grid_x = w+1;
+    df::global::init->display.grid_y = h+1;
+    *window_x = x;
+    *window_y = y;
+    *window_z = z;
+    //force full redraw
+    df::global::gps->force_full_display_count = 1;
+    for (int ty = 0; ty < h; ty++)
+    for (int tx = 0; tx < w; tx++)
+    {
+        for (int i = 0; i < 4; i++)
+        {
+            int t = (tx + 1)*win_h + ty + 1;
+            s[t * 4 + i] = 0;
+        }
+    }
+    render_map();
+    //restore state
+    *window_x = was_x;
+    *window_y = was_y;
+    *window_z = was_z;
+    df::global::init->display.grid_x = gx;
+    df::global::init->display.grid_y = gy;
+
+    lua_createtable(L,w*h*4,0);
```

Comment:
I think a `PenArray` would work well here, but I'm not sure if there's currently an API to create those outside of LuaApi.cpp.

Reply:
I forget how it works, but there was a system in place to push c++ stuff into lua.